File: others/2100-2200/2141. 同时运行 N 台电脑的最长时间.cpp
//
// Created by 123 on 25-5-22.
//
#include "bits/stdc++.h"
using namespace std;
using ll = long long;
constexpr int MOD = 1'000'000'007;

class Solution {
public:
  long long maxRunTime(int n, vector<int>& batteries) {
    ll sum = reduce(batteries.begin(), batteries.end(), 0LL);
    ll l = 0;
    ll r = sum / n + 1e5;
    ll ans = 0;
    while (l < r) {
      ll mid = l + (r - l) / 2;
      ll temp = 0;
      for (ll b : batteries) {
        temp += min(b, mid);
      }
      if (temp >= n * mid * 1LL) {
        ans = mid;
        l = mid + 1;
      } else {
        r = mid;
      }
    }
    return ans;
  }
};

File: leetcode/2100-2200/2141. 同时运行 N 台电脑的最长时间.py
from collections import *
from functools import *
from math import *
from itertools import *
from heapq import *
from bisect import *
from typing import *
from queue import *
from string import *

DIRS = [(1, 1), (1, -1), (-1, -1), (-1, 1)]
Max = lambda a, b: b if b > a else a
Min = lambda a, b: b if b < a else a
INF = float('inf')
MOD = int(1e9 + 7)


class Solution:
    def maxRunTime(self, n: int, batteries: List[int]) -> int:
        l, r = 0, sum(batteries) // n + 1
        while l + 1 < r:
            x = (l + r) // 2
            if n * x <= sum(min(b, x) for b in batteries):
                l = x
            else:
                r = x
        return l


class Solution:
    def maxRunTime(self, n: int, batteries: List[int]) -> int:
        batteries.sort(reverse=True)
        s = sum(batteries)
        for b in batteries:
            if b <= s // n:
                return s // n
            s -= b
            n -= 1
